hutch_python/tests/test_namespace.py

```diff
@@ -66,9 +66,31 @@ def test_tree_namespace():
     assert xpp.sb2.obj5 == 5
 
 
-def test_conflicting_name():
-    logger.debug('test_conflicting_name')
+def test_same_name():
+    logger.debug('test_same_name')
     # This should be ok, but make sure the warning is covered
     scope = SimpleNamespace(hutch_stand=SimpleNamespace(dev=1),
                             hutch_stand_dev=2)
     tree_namespace(scope=scope)
+
+
+def test_conflicting_name():
+    logger.debug('test_conflicting_name')
+    # There was a bug where names could conflict with eachother
+    scope = SimpleNamespace(daq=Signal(name='daq'),
+                            daq_repeat_echo='daq_repeat_echo',
+                            daq_repeat=Signal(name='daq_repeat'),
+                            daq_deep='daq_deep',
+                            daq_deep_deep=Signal(name='daq_deep_deep'))
+    ns = tree_namespace(scope=scope)
+    assert ns.daq.name == 'daq'
+    assert ns.daq.repeat.echo == 'daq_repeat_echo'
+    assert ns.daq.repeat.name == 'daq_repeat'
+
+
+def test_invalid_names():
+    logger.debug('test_invalid_names')
+    # Do not crash, but cover warnings and verify not in tree
+    scope = SimpleNamespace(cheese_in_wheel=4, why_1=1)
```

Comment:
Wait is the behavior we want? In my mind both these objects should be in the tree as:

```python
cheese.in_wheel
why_1
```

Reply:
Fair point, I'll look at it.

Reply:
After some thought, we'd have to make some rules about when to merge the invalid sections with the prior section vs the following section. The example from the test is complicated when you add additional objects e.g.:
`cheese_in`, `cheese`, `cheese_in_wheel_1`
So how do you sort these into the tree?
`cheese.in_wheel`
`cheese_in`
`cheese`
`cheese.in_wheel_1`
It falls apart, even if you have strict rules e.g. "always merge with the previous key"... you have to ask the user to change their names or accept a malformed tree.

If we're ok with a malformed tree, I will enact the "always merge with previous key" rule, so these become:
`cheese_in.wheel`
`cheese_in`
`cheese`
`cheese_in.wheel_1`
The "malformed" here is that there are a lot of `cheese` names that don't sort under `cheese`

Reply:
Agreed. I think not having them might still be better then malforming them as long as they are accessible. That being said, less concerned about this if we just limit to specific groups. 

What's wrong with merging to next key then back if not available? I think that gives you:

`cheese.in_wheel`
`cheese`
`cheese_in`
`cheese.in_wheel_1`

which at least gives you three of the four and `cheese_in` will never be "splittable" because "in" will never be valid.

Reply:
The algorithm in my head for "next, else back" is much uglier than the one for "always previous", which is much uglier than the code as implemented. I'm a sucker for keeping code readable.